libs/exponential-histogram/src/main/java/org/elasticsearch/exponentialhistogram/ZeroBucket.java

```diff
@@ -0,0 +1,154 @@
+
+    // A singleton for an empty zero bucket with the smallest possible threshold.
+    private static final ZeroBucket MINIMAL_EMPTY = new ZeroBucket(MIN_INDEX, MIN_SCALE, 0);
+
+    /**
+     * Creates a new zero bucket with a specific threshold and count.
+     *
+     * @param zeroThreshold The threshold defining the bucket's range [-zeroThreshold, +zeroThreshold].
+     * @param count         The number of values in the bucket.
+     */
+    public ZeroBucket(double zeroThreshold, long count) {
+        this(computeIndex(zeroThreshold, MAX_SCALE) + 1, MAX_SCALE, count);
+    }
+
+    /**
+     * @return A singleton instance of an empty zero bucket with the smallest possible threshold.
+     */
+    public static ZeroBucket minimalEmpty() {
+        return MINIMAL_EMPTY;
+    }
+
+    /**
+     * Creates a zero bucket with the smallest possible threshold and a given count.
+     *
+     * @param count The number of values in the bucket.
+     * @return A new {@link ZeroBucket}.
+     */
+    public static ZeroBucket minimalWithCount(long count) {
+        if (count == 0) {
+            return MINIMAL_EMPTY;
+        } else {
+            return new ZeroBucket(MINIMAL_EMPTY.index, MINIMAL_EMPTY.scale(), count);
+        }
+    }
+
+    /**
+     * Merges this zero bucket with another one.
+     * <ul>
+     *     <li>If the other zero bucket is empty, this instance is returned unchanged.</li>
+     *     <li>Otherwise, the zero threshold is increased if necessary (by taking the maximum of the two), and the counts are summed.</li>
+     * </ul>
+     *
+     * @param other The other zero bucket to merge with.
+     * @return A new {@link ZeroBucket} representing the merged result.
+     */
+    public ZeroBucket merge(ZeroBucket other) {
+        if (other.count == 0) {
+            return this;
+        } else {
+            long totalCount = count + other.count;
+            // Both are populated, so we need to use the higher zero-threshold.
+            if (this.compareZeroThreshold(other) >= 0) {
+                return new ZeroBucket(index, scale, totalCount);
+            } else {
+                return new ZeroBucket(other.index, other.scale, totalCount);
+            }
+        }
+    }
+
+    /**
+     * Collapses all buckets from the given iterators whose lower boundaries are smaller than the zero threshold.
+     * The iterators are advanced to point at the first, non-collapsed bucket.
+     *
+     * @param bucketIterators The iterators whose buckets may be collapsed.
+     * @return A potentially updated {@link ZeroBucket} with the collapsed buckets' counts and an adjusted threshold.
+     */
+    public ZeroBucket collapseOverlappingBuckets(BucketIterator... bucketIterators) {
+        ZeroBucket current = this;
+        ZeroBucket previous;
+        do {
+            previous = current;
+            for (BucketIterator buckets : bucketIterators) {
+                current = current.collapseOverlappingBuckets(buckets);
```

Comment:
Nit - using the same name here makes this look recursive at first glance.  The fact that the two methods have a bunch of code in between them isn't helping. Maybe rename this to `collapsSingleOverlappingBucket` or something like that?

Reply:
renamed this method to `collapseOverlappingBucketsForAll`. `collapsSingleOverlappingBucket` would be misleading because that method can still collapse multiple buckets, but does so only from a single iterator.